File: backend/src/scrapling.js
const clean=(value,max=1000)=>String(value??'').replace(/\s+/g,' ').trim().slice(0,max);

export function scraplingConfigured(env={}){
  const raw=clean(env.SCRAPLING_SERVICE_URL||'',500);if(!raw)return false;
  try{const url=new URL(raw);return url.protocol==='https:'&&Boolean(url.hostname);}catch{return false;}
}

export async function fetchWithScrapling(env={},targetUrl){
  if(!scraplingConfigured(env))throw new Error('Scrapling service is not configured');
  const serviceUrl=new URL(clean(env.SCRAPLING_SERVICE_URL,500));
  const target=new URL(String(targetUrl||''));
  if(!['http:','https:'].includes(target.protocol))throw new Error('A public http/https URL is required');
  const headers={'Content-Type':'application/json',Accept:'application/json'};
  const token=String(env.SCRAPLING_SERVICE_TOKEN||'').trim();if(token)headers.Authorization=`Bearer ${token}`;
  let response;
  try{
    response=await fetch(serviceUrl.href,{method:'POST',headers,body:JSON.stringify({url:target.href}),signal:AbortSignal.timeout(35000)});
  }catch(cause){throw new Error(`Scrapling request failed: ${String(cause?.message||cause).slice(0,160)}`);}
  const payload=await response.json().catch(()=>null);
  if(!response.ok)throw new Error(clean(payload?.detail||payload?.error||`Scrapling returned ${response.status}`,180));
  const markdown=String(payload?.data?.markdown||'').trim();const metadata=payload?.data?.metadata||{};
  if(payload?.success!==true||!markdown||metadata.source!=='scrapling-fallback')throw new Error('Invalid Scrapling response');
  let sourceUrl='';try{sourceUrl=new URL(String(metadata.sourceURL||metadata.url||target.href)).href;}catch{throw new Error('Invalid Scrapling response');}
  return {success:true,data:{markdown:markdown.slice(0,60000),metadata:{...metadata,title:clean(metadata.title||new URL(sourceUrl).hostname,180),sourceURL:sourceUrl,url:sourceUrl,statusCode:Number(metadata.statusCode)||200,source:'scrapling-fallback',fetchedAt:clean(metadata.fetchedAt||new Date().toISOString(),80)}}};
}

File: customer/api/scrapling.py
import ipaddress
import os
import socket
from datetime import datetime, timezone
from urllib.parse import urlparse

from fastapi import FastAPI, Header, HTTPException
from pydantic import BaseModel
from scrapling.fetchers import Fetcher

app = FastAPI()
MAX_TEXT_CHARS = 60000


class ScrapeRequest(BaseModel):
    url: str


def _authorized(authorization: str | None) -> bool:
    token = os.getenv("SCRAPLING_SERVICE_TOKEN", "").strip()
    if not token:
        return True
    return authorization == f"Bearer {token}"


def _public_url(value: str) -> str:
    parsed = urlparse(str(value or "").strip())
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise HTTPException(status_code=400, detail="A valid public http/https URL is required")
    host = parsed.hostname.rstrip(".").lower()
    if host in {"localhost"} or host.endswith((".localhost", ".local", ".internal", ".lan")):
        raise HTTPException(status_code=400, detail="Private or local URLs are not allowed")
    try:
        addresses = {
            row[4][0]
            for row in socket.getaddrinfo(
                host,
                parsed.port or (443 if parsed.scheme == "https" else 80),
                type=socket.SOCK_STREAM,
            )
        }
    except socket.gaierror as exc:
        raise HTTPException(status_code=422, detail="Unable to resolve target host") from exc
    for raw in addresses:
        try:
            ip = ipaddress.ip_address(raw)
        except ValueError:
            continue
        if not ip.is_global:
            raise HTTPException(status_code=400, detail="Private or non-global target addresses are not allowed")
    return parsed.geturl()


def _title(page) -> str:
    try:
        value = page.css("title::text").get()
        return str(value or "").strip()[:180]
    except Exception:
        return ""


@app.get("/api/scrapling")
def health():
    return {"status": "ok", "service": "leadintel-scrapling"}


@app.post("/api/scrapling")
def scrape(payload: ScrapeRequest, authorization: str | None = Header(default=None)):
    if not _authorized(authorization):
        raise HTTPException(status_code=401, detail="Unauthorized")
    url = _public_url(payload.url)
    try:
        page = Fetcher.get(url, stealthy_headers=True, impersonate="chrome", timeout=30000)
        text = str(page.get_all_text(separator="\n", strip=True) or "").strip()
        if not text:
            raise HTTPException(status_code=422, detail="No readable page content was returned")
        status = int(getattr(page, "status", 200) or 200)
        final_url = str(getattr(page, "url", url) or url)
        return {
            "success": True,
            "data": {
                "markdown": text[:MAX_TEXT_CHARS],
                "metadata": {
                    "title": _title(page) or urlparse(final_url).hostname,
                    "sourceURL": final_url,
                    "url": final_url,
                    "statusCode": status,
                    "source": "scrapling-fallback",
                    "fetchedAt": datetime.now(timezone.utc).isoformat(),
                },
            },
        }
    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Scrapling extraction failed: {str(exc)[:160]}") from exc
